### PR TITLE
[Doc] Add dataset & dataloader tutorial

### PR DESCRIPTION
**Pull request by C1rN09**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: docs/zh_cn/tutorials/dataset.md [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+        sampler=dict(
+            type='DefaultSampler',
+            shuffle=True),
+        dataset=torchvision.datasets.CIFAR10(...),
+        collate_fn=dict(type='default_collate')
+    )
+)
+```
+
+在这种情况下，数据加载器会在实际被用到时、在执行器内部被构建。
```
Comment (zhouzaida) [inline docs/zh_cn/tutorials/dataset.md:36]:
```suggestion
在这种情况下，数据加载器会在实际被用到时，在执行器内部被构建。
```

File: docs/zh_cn/tutorials/dataset.md [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+- `shuffle` 设置为 `True` 时会打乱数据集的读取顺序
+- `seed` 打乱数据集所用的随机种子，通常不需要在此手动设置，会从 `Runner` 的 `randomness` 入参中读取
+- `round_up` 设置为 `True` 时，与 PyTorch `DataLoader` 中设置 `drop_last=False` 行为一致。如果你在迁移 PyTorch 的项目，你可能需要注意这一点。
+
+```{note}
+更多关于 `DefaultSampler` 的内容可以参考 [API 文档](mmengine.dataset.DefaultSampler)
+```
+
+`DefaultSampler` 适用于绝大部分情况，并且我们保证在执行器中使用它时，随机数等容易出错的细节都被正确地处理，防止你陷入多进程训练的常见陷阱。如果你想要使用基于迭代次数(iteration-based)的训练流程，你也许会对 [InfiniteSampler](mmengine.dataset.InfiniteSampler) 感兴趣。如果你有更多的进阶需求，你可能会想要参考上述两个内置 `sampler` 的代码，实现一个自定义的 `sampler` 并注册到 `DATA_SAMPLERS` 根注册器中。
```
Comment (zhouzaida) [inline docs/zh_cn/tutorials/dataset.md:88]:
```suggestion
`DefaultSampler` 适用于绝大部分情况，并且我们保证在执行器中使用它时，随机数等容易出错的细节都被正确地处理，防止你陷入多进程训练的常见陷阱。如果你想要使用基于迭代次数（iteration-based）的训练流程，你也许会对 [InfiniteSampler](mmengine.dataset.InfiniteSampler) 感兴趣。如果你有更多的进阶需求，你可能会想要参考上述两个内置 `sampler` 的代码，实现一个自定义的 `sampler` 并注册到 `DATA_SAMPLERS` 根注册器中。
```

File: docs/zh_cn/tutorials/dataset.md [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    train_dataloader=dict(
+        sampler=dict(type='MySampler'),
+        ...
+    )
+)
+```
+
+### 不起眼的 collate_fn
+
+PyTorch 的 `DataLoader` 中，`collate_fn` 这一参数常常被使用者们忽略，但在 MMEngine 中你需要额外注意：当你传入 `dict` 来构造数据加载器时，MMEngine 会默认使用内置的 [pseudo_collate](mmengine.dataset.pseudo_collate)，这一点明显区别于 PyTorch 默认的 [default_collate](https://pytorch.org/docs/stable/data.html#torch.utils.data.default_collate)。因此，当你迁移 PyTorch 项目时，需要在配置文件中手动指明 `collate_fn` 以保持行为一致。
```
Comment (zhouzaida) [inline docs/zh_cn/tutorials/dataset.md:105]:
```suggestion
PyTorch 的 `DataLoader` 中，`collate_fn` 这一参数常常被使用者忽略，但在 MMEngine 中你需要额外注意：当你传入 `dict` 来构造数据加载器时，MMEngine 会默认使用内置的 [pseudo_collate](mmengine.dataset.pseudo_collate)，这一点明显区别于 PyTorch 默认的 [default_collate](https://pytorch.org/docs/stable/data.html#torch.utils.data.default_collate)。因此，当你迁移 PyTorch 项目时，需要在配置文件中手动指明 `collate_fn` 以保持行为一致。
```

File: docs/zh_cn/tutorials/dataset.md [hunk 4]
```diff
@@ -0,0 +1,198 @@
+# 数据集（Dataset）与数据加载器（DataLoader）
+
+```{hint}
+如果你没有接触过 PyTorch 的数据集与数据加载器，我们推荐先浏览 [PyTorch 官方教程](https://pytorch.org/tutorials/beginner/basics/data_tutorial.html)以了解一些基本概念
+```
+
+数据集与数据加载器是 MMEngine 中训练流程的必要组件，它们的概念来源于 PyTorch，并且在含义上与 PyTorch 保持一致。通常来说，数据集定义了数据的总体数量、读取方式以及预处理，而数据加载器则在不同的设置下（如 `batch_size`、随机乱序、多进程）迭代地加载数据。两者共同构成了数据源。在本篇教程中，我们将按照从外（数据加载器）到内（数据集）的顺序，逐步介绍它们在 MMEngine 执行器中的用法，并给出一些常用示例。读完本篇教程，你将会：
```
Comment (HAOCHENYE) [inline docs/zh_cn/tutorials/dataset.md:7]:
```suggestion
数据集与数据加载器是 MMEngine 中训练流程的必要组件，它们的概念来源于 PyTorch，并且在含义上与 PyTorch 保持一致。通常来说，数据集定义了数据的总体数量、读取方式以及预处理，而数据加载器则在不同的设置下（如批次大小（`batch_size`）、随机乱序（`shuffle`）、多进程（`num_workers`））迭代地加载数据。前者经后者封装后构成了数据源。在本篇教程中，我们将按照从外（数据加载器）到内（数据集）的顺序，逐步介绍它们在 MMEngine 执行器中的用法，并给出一些常用示例。读完本篇教程，你将会：
```

File: docs/zh_cn/tutorials/dataset.md [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+## 数据加载器详解
+
+在执行器（`Runner`）中，你可以分别配置以下 3 个参数来指定对应的数据加载器
+
+- `train_dataloader`：在 `Runner.train()` 中被使用，为模型提供训练数据
+- `val_dataloader`：在 `Runner.val()` 中被使用，也会在 `Runner.train()` 中每间隔一段时间被使用，用于模型的验证评测
+- `test_dataloader`：在 `Runner.test()` 中被使用，用于模型的测试
+
+MMEngine 完全支持 PyTorch 的原生 `DataLoader`，因此上述 3 个参数均可以直接传入构建好的 `DataLoader`，如[15分钟上手](../get_started/15_minutes.md)中的例子所示。同时，借助 MMEngine 的[注册机制](../advanced_tutorials/registry.md)，以上参数也可以传入 `dict`，如下面代码（以下简称例 1）所示。
```
Comment (HAOCHENYE) [inline docs/zh_cn/tutorials/dataset.md:21]:
```suggestion
MMEngine 完全支持 PyTorch 的原生 `DataLoader`，因此上述 3 个参数均可以直接传入构建好的 `DataLoader`，如[15分钟上手](../get_started/15_minutes.md)中的例子所示。同时，借助 MMEngine 的[注册机制](../advanced_tutorials/registry.md)，以上参数也可以传入 `dict`，如下面代码（以下简称例 1）所示，字典中配置的键值对完全对标 `Dataloader` 的构造参数 
```

File: docs/zh_cn/tutorials/dataset.md [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+sampler = DefaultSampler(dataset, shuffle=True)
+
+runner = Runner(
+    train_dataloader=dict(
+        batch_size=32,
+        sampler=sampler,
+        dataset=dataset,
+        collate_fn=dict(type='default_collate')
+    )
+)
```
Comment (HAOCHENYE) [inline docs/zh_cn/tutorials/dataset.md:71]: 这里 train_dataloader 还需要是字典么

File: docs/zh_cn/tutorials/dataset.md [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+                dict(type='RandomHorizontalFlip'),
+                dict(type='ToTensor'),
+                dict(type='Normalize', **norm_cfg)])
+    )
+)
+```
+
+```{note}
+上述例子中大量使用了[注册机制](../advanced_tutorials/registry.md)，并且用到了 MMEngine 中的 [Compose](mmengine.dataset.Compose)。如果你急需在配置文件中使用 `torchvision` 数据集，你可以参考上述代码并略作修改。但我们更加推荐你有需要时在下游库（如 [mmdet](https://github.com/open-mmlab/mmdetection) 和 [mmcls](https://github.com/open-mmlab/mmclassification) 等）中寻找对应的数据集实现，从而获得更好的使用体验。
+```
```
Comment (HAOCHENYE) [inline docs/zh_cn/tutorials/dataset.md:188]:
```suggestion
上述例子中大量使用了[注册机制](../advanced_tutorials/registry.md)，并且用到了 MMEngine 中的 [Compose](mmengine.dataset.Compose)。如果你急需在配置文件中使用 `torchvision` 数据集，你可以参考上述代码并略作修改。但我们更加推荐你有需要时在下游库（如 [MMDet](https://github.com/open-mmlab/mmdetection) 和 [MMCls](https://github.com/open-mmlab/mmclassification) 等）中寻找对应的数据集实现，从而获得更好的使用体验。
```
```

File: docs/zh_cn/tutorials/dataset.md [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+
+### 自定义数据集
+
+你可以像使用 PyTorch 一样，自由地定义自己的数据集，或将之前 PyTorch 项目中的数据集拷贝过来。如果你想要了解如何自定义数据集，你可以参考 [PyTorch 官方教程](https://pytorch.org/tutorials/beginner/basics/data_tutorial.html#creating-a-custom-dataset-for-your-files)
```
Comment (HAOCHENYE) [inline docs/zh_cn/tutorials/dataset.md:192]:
```suggestion
你可以像使用 PyTorch 一样，自由地定义自己的数据集，或将之前 PyTorch 项目中的数据集拷贝过来。如果你想要了解如何自定义数据集，可以参考 [PyTorch 官方教程](https://pytorch.org/tutorials/beginner/basics/data_tutorial.html#creating-a-custom-dataset-for-your-files)
```

**Review by zhouzaida — APPROVED** [review]

**Review by HAOCHENYE — APPROVED** [review]
Approved!!!
image
